x/datapool/keeper/pool.go

```diff
@@ -553,3 +544,129 @@ func (k Keeper) increaseNumIssuedNFT(ctx sdk.Context, pool *types.Pool) {
+	pool, err := k.GetPool(ctx, redeemNFT.PoolId)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	if pool.GetRound() != redeemNFT.Round {
+		return nil, types.ErrRoundNotMatched
+	}
+
+	nftContractAcc, err := sdk.AccAddressFromBech32(pool.NftContractAddr)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress
+	}
+
+	moduleAddr := types.GetModuleAddress()
+	transferNFTMsg := types.NewTransferNFTMsg(moduleAddr.String(), strconv.FormatUint(redeemNFT.NftId, 10))
+
+	transferNFTMsgBz, err := json.Marshal(transferNFTMsg)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	redeemerAcc, err := sdk.AccAddressFromBech32(redeemNFT.Redeemer)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	redeemTokenId, err := k.GetRedeemerDataPass(ctx, pool.PoolId, redeemerAcc)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	if !contains(redeemTokenId, strconv.FormatUint(redeemNFT.NftId, 10)) {
+		return nil, types.ErrInvalidTokenId
+	}
+
+	zeroFund, err := sdk.ParseCoinsNormalized("0umed")
+	if err != nil {
+		return nil, sdkerrors.Wrapf(err, "error in parsing coin")
+	}
+
+	_, err = k.wasmKeeper.Execute(ctx, nftContractAcc, redeemerAcc, transferNFTMsgBz, zeroFund)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	nftRedeemReceipt := types.NewDataAccessNFTRedeemReceipt(redeemNFT.PoolId, redeemNFT.Round, redeemNFT.NftId, uint64(ctx.BlockHeight()), redeemNFT.Redeemer)
+
+	err = k.SetDataPassRedeemReceipt(ctx, *nftRedeemReceipt)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	return nftRedeemReceipt, nil
+}
+
+func (k Keeper) GetDataPassRedeemReceipt(ctx sdk.Context, poolID, round uint64, redeemer sdk.AccAddress) (types.DataPassRedeemReceipt, error) {
+	key := types.GetKeyPrefixNFTRedeemReceipt(poolID, round, redeemer)
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(key)
+
+	var receipt types.DataPassRedeemReceipt
+	err := k.cdc.UnmarshalBinaryLengthPrefixed(bz, &receipt)
+	if err != nil {
+		return types.DataPassRedeemReceipt{}, sdkerrors.Wrap(types.ErrGetDataPassRedeemReceipt, err.Error())
+	}
+
+	return receipt, nil
+}
+
+func (k Keeper) SetDataPassRedeemReceipt(ctx sdk.Context, redeemReceipt types.DataPassRedeemReceipt) error {
+	redeemer, err := sdk.AccAddressFromBech32(redeemReceipt.Redeemer)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, err.Error())
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	receiptKey := types.GetKeyPrefixNFTRedeemReceipt(redeemReceipt.PoolId, redeemReceipt.Round, redeemer)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(&redeemReceipt)
+	store.Set(receiptKey, bz)
+
+	return nil
+}
+
+func (k Keeper) GetRedeemerDataPass(ctx sdk.Context, poolID uint64, redeemer sdk.AccAddress) ([]string, error) {
```

Comment:
The name of this keeper function is confusing. I know it's the hardest thing among what we do 😢 
I suggest a new name.
```suggestion
func (k Keeper) GetAllDataPassByAddr(ctx sdk.Context, poolID uint64, redeemer sdk.AccAddress) ([]string, error) {
```